### ISSUE TITLE
Deployments stalling when using "parrallel" keyword

### ISSUE THREAD

**Issue by dkennedy09** [onicagroup/runway#509]
**Describe the bug**
We are running into an issue in runway version 1.15.0 where deployments will stall on the following log message when trying to use `parallel` or `parallel_regions` to deploy to us-east-1 and us-west-2. This always results in the deployment timing out unless it is manually aborted.
```
NOTICE:runway.core.components.deployment:deployment_1:processing deployment (in progress)
VERBOSE:runway.core.components.deployment:deployment_1:attempting to deploy to region(s): us-east-1, us-west-2
INFO:runway.core.components.deployment:deployment_1:processing regions in parallel... (output will be interwoven)
```
Does anyone know why this might be happening?

**Comment by ITProKyle** [comment]
Have you checked the debug logs to get more info on where it's getting stuck?

Is it able to run when not using parallel?

Have you tried using the latest version of runway?

What does your config file look like?

What version of python are you using?

> This always results in the deployment timing out unless it is manually aborted.

What is timing out? Is it your CI/CD pipeline or is runway outputting an error when it times out? If runway is timing out and outputting an error, what is the full error (including traceback)?

**Comment by gogineni99** [comment]
Facing the same issue. Tried with runway 1.18.1
Parallel works fine with CloudFormation, errors when using with Terraform
Python version : 3.8.5

Terraform module works fine without parallel.

Running Terraform on the machine directly throws this error:

**Comment by gogineni99** [comment]
Running `runway deploy` in Docker container throws this error

`NOTICE:runway.core.components.deployment:deployment_1:processing deployment (in progress)
VERBOSE:runway.core.components.deployment:deployment_1:attempting to deploy to region(s): us-east-1, us-west-2
INFO:runway.core.components.deployment:deployment_1:processing regions in parallel... (output will be interwoven)
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/queues.py", line 234, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: can't pickle _thread.RLock objects
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/queues.py", line 234, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: can't pickle _thread.RLock objects`

**Comment by ITProKyle** [comment]
Currently, CloudFormation modules are the only modules known and guaranteed to work correctly with parallelism. YMMV for other module types due to their reliance on consuming and modifying local files and directories (`.terraform`, `.serverless`, etc) at runtime which is beyond our control.

```
OSError: Unable to trash file /home/gogineni/sandbox/terraform-parallel/runway/parallel-test-tf/.terraform/modules: No such file or directory
```

This error is unavoidable at the moment due to our use of `send2trash` which the [last attempt to remove, was denied](https://github.com/onicagroup/runway/pull/412#pullrequestreview-467858238). The conversation around removing the use of `send2trash` was brought up again last week and the stance was still the same.

```
  File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: can't pickle _thread.RLock objects
```

This is likely due to the release of python 3.6 being used. I have seen it occur in some of the older releases.

**Comment by gogineni99** [comment]
Understood. Update to docs mentioning that parallel_regions works only for CloudFormation would help users.

**Comment by troyready** [comment]
I agree, thanks @gogineni99 .

We should go a step farther and just disable parallel execution for non-CFN module types. In the future (or same PR) it could be re-enabled with an option to run TF/SLS/CDK modules in their own temporary directories simultaneously.

**Comment by ITProKyle** [comment]
I encountered this recently with CloudFormation. However it did not _stall_, it deployed successfully and displayed all the expected output (although delayed).

I believe that the output is being delayed somewhere along the line and only being output once the process exits but will have to dig into it further to find where this could be happening.

**Comment by ITProKyle** [comment]
I did a bit more digging into the root cause of the output being delayed during parallel execution. There are a few factors contributing to it that I will try to explain in detail with a tl;dr at the end for those that don't care about the inner workings of python.

Prior to python 3.7, when creating a subprocess on a POSIX it would **fork** the parent process (using functionality built into the OS). This copies everything that is currently in memory to each subprocess such as logging configuration. However, Windows does not have the same functionality so each subprocess **spawns** a new instance of python which does not carry over anything from the parent process. Starting with python 3.7, the default action when creating a new subprocess on Windows and macOS is to **spawn** a new process while other POSIX systems _should_ still default to **fork**.

In Runway 1.10.0 we restructured some of the code to consolidate some code that was inherited from another codebase into the Runway code base. Prior to this, logging was being setup for CloudFormation modules each time a module was run. While inefficient, it hid the above mentioned **fork** vs **spawn** differences.

As for solving the issue, there are a few things to consider:

1. Reverting the changes made to 1.10.0, we would lose quite a bit of functionality. IMO, this is not an option.
2. There is an option to specifically tell python use **fork** when creating new process. However, this still does not support Windows and requires `python >= 3.7`. Since Runway 1.x supports `python < 3.7`, it can't be used until the next major release. (`mp_context` arg [[source]](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor))
3. When implementing the solution in the next major release, we could force all child processes to be **spawn** rather than **fork** to handle them all uniformly. However, this would break our binary releases. [[source]](https://docs.python.org/3.9/library/multiprocessing.html#contexts-and-start-methods)

The most viable workaround (for now) would be to force all child processes to use **fork** (unless on Windows) in the next major release since we are already dropping support for `python < 3.7` as part of it. This should solve for most cases, except for Windows users who will continue to have the same delayed output.

With the workaround in place, we can put focus on improving our support of multiprocessing/multithreading to devise a long term solution that would solve for all platforms (#298). While forking processes on POSX _works_ it has its draw backs especially when trying to do more complex multiprocess actions.

# tl;dr

- Runway 1.10.0 moved some code that was hiding this issue with python multiprocessing and how we are trying to use it
- `python >= 3.7` changes some defaults for multiprocessing making the above point more prevalent on newer versions of python
